yt_dlp/extractor/common.py

```diff
@@ -2536,26 +2545,79 @@ def _extract_mpd_formats_and_subtitles(
         mpd_url = urlh.url
         mpd_base_url = base_url(mpd_url)
 
-        return self._parse_mpd_formats_and_subtitles(
-            mpd_doc, mpd_id, mpd_base_url, mpd_url)
+        return self._parse_mpd_periods(mpd_doc, mpd_id, mpd_base_url, mpd_url)
 
     def _parse_mpd_formats(self, *args, **kwargs):
         fmts, subs = self._parse_mpd_formats_and_subtitles(*args, **kwargs)
         if subs:
             self._report_ignoring_subs('DASH')
+        if any(f.get('is_dash_periods') for f in fmts):
+            self.report_warning(bug_reports_message(
+                'Ignoring multiple periods found in the DASH manifest; '
+                'if part of the video is missing,'
+            ), only_once=True)
         return fmts
 
-    def _parse_mpd_formats_and_subtitles(
-            self, mpd_doc, mpd_id=None, mpd_base_url='', mpd_url=None):
+    def _parse_mpd_formats_and_subtitles(self, *args, **kwargs):
+        periods = self._parse_mpd_periods(*args, **kwargs)
+        return self._merge_mpd_periods(periods)
+
+    def _merge_mpd_periods(self, periods):
+        """
+        Combine all formats and subtitles from an MPD manifest into a single list,
+        by concatenate streams with similar formats.
+        """
+        formats, subtitles = [], {}
+
+        # merge periods by concatenating fragments
+        # require almost all parameters to be the same
+        shared_keys = [k for k in periods[0]['formats'][0].keys()
+                       if k not in ('format_id', 'fragments', 'manifest_stream_number',
+                                    'dash_period_id', 'dash_period_idx')]
+
+        # create merged formats
+        stream_numbers = collections.defaultdict(int)
+        groups = {}
+        formats = []
+        for period in periods:
+            for f in period['formats']:
+                format_key = tuple(f[k] for k in shared_keys)
+                if 'fragments' not in f or format_key not in groups:
+                    # unmergeable format, or first period of this format
+                    mf = {}
+                    mf.update(f)
+                    mf['manifest_stream_number'] = stream_numbers[mf['url']]
+                    formats.append(mf)
+                    # keep track of mergeable formats
+                    if 'fragments' in f:
+                        mf['is_dash_periods'] = True
+                        mf['fragments'] = f['fragments'].copy()
+                        groups[format_key] = mf
+                else:
+                    # follow-up period of a mergeable format
+                    groups[format_key]['fragments'] += f['fragments']
+
+        for period in periods:
+            subtitles.update(period['subtitles'])
+
+        if len(subtitles) > 0 and len(periods) > 1:
+            self.report_warning(bug_reports_message(
+                'Ignoring multiple periods found in the DASH manifest; '
+                'if part of the subtitles are missing,'
+            ), only_once=True)
```

Comment:
pls explain

Reply:
The subtitles are now appended to the list, so the subtitles for multiple periods will no longer be overwritten. I have removed the warning.

Maybe the warning should be put back though.

The reason this warning was here is that the combination of multi-period streams with subtitles might not yet work. For the audio/video streams, the fragments lists are concatenated to download one big stream. For subtitles, there is currently no such thing: they will show up as separate files.

I have not yet found a streaming URL with multi-period subtitles where I can test this, so I have no idea what will happen. Some MPD manifests seem to list the same subtitle file in each period and rely on the time codes to work over multiple periods. In the current implementation, I think the single subtitle file might work with the concatenated video. But I am not entirely sure, hence the warning.

If the subtitles are split over multiple files, then I think you'd have to download all of them, and I suspect you  may have to correct the timestamps as well. This is speculation, I have no examples.

Maybe it would be good to keep the warning that multi-period streams with subtitles are untested, so people can submit bug reports and examples to find these edge cases in the wild?

Reply:
The warning should be when there is "subtitle split over multiple periods", correct? That is not whatthe current code is checking

Reply:
Correct. See a better alternative below.